boa_engine/src/module/source.rs

```diff
@@ -0,0 +1,1903 @@
+        LocalExportEntry, ReExportImportName,
+    },
+    operations::{
+        bound_names, contains, lexically_scoped_declarations, var_scoped_declarations,
+        ContainsSymbol,
+    },
+    Declaration, ModuleItemList,
+};
+use boa_gc::{custom_trace, empty_trace, Finalize, Gc, GcRefCell, Trace};
+use boa_interner::Sym;
+use rustc_hash::{FxHashMap, FxHashSet};
+
+use crate::{
+    builtins::{promise::PromiseCapability, Promise},
+    bytecompiler::{ByteCompiler, NodeKind},
+    environments::{BindingLocator, CompileTimeEnvironment, EnvironmentStack},
+    module::ModuleKind,
+    object::{FunctionObjectBuilder, JsPromise, RecursionLimiter},
+    realm::Realm,
+    vm::{CallFrame, CodeBlock, CompletionRecord, Opcode},
+    Context, JsArgs, JsError, JsNativeError, JsObject, JsResult, JsString, JsValue, NativeFunction,
+};
+
+use super::{
+    BindingName, GraphLoadingState, Module, Referrer, ResolveExportError, ResolvedBinding,
+};
+
+/// Information for the [**Depth-first search**] algorithm used in the
+/// [`Module::link`] and [`Module::evaluate`] methods.
+#[derive(Clone, Copy, Debug, Finalize)]
+pub(super) struct DfsInfo {
+    dfs_index: usize,
+    dfs_ancestor_index: usize,
+}
+
+// SAFETY: `DfsInfo` only contains primitive types, making this safe.
+unsafe impl Trace for DfsInfo {
+    empty_trace!();
+}
+
+/// Current status of a [`SourceTextModule`].
+///
+/// Roughly corresponds to the `[[Status]]` field of [**Cyclic Module Records**][cyclic],
+/// but with a state machine-like design for better correctness.
+///
+/// [cyclic]: https://tc39.es/ecma262/#table-cyclic-module-fields
+#[derive(Debug, Finalize, Default)]
+enum Status {
+    #[default]
+    Unlinked,
+    Linking {
+        info: DfsInfo,
+    },
+    PreLinked {
+        context: SourceTextContext,
+        info: DfsInfo,
+    },
+    Linked {
+        context: SourceTextContext,
+        info: DfsInfo,
+    },
+    Evaluating {
+        context: SourceTextContext,
+        top_level_capability: Option<PromiseCapability>,
+        cycle_root: SourceTextModule,
+        info: DfsInfo,
+        async_eval_index: Option<usize>,
+    },
+    EvaluatingAsync {
+        context: SourceTextContext,
+        top_level_capability: Option<PromiseCapability>,
+        cycle_root: SourceTextModule,
+        async_eval_index: usize,
+        pending_async_dependencies: usize,
+    },
+    Evaluated {
+        top_level_capability: Option<PromiseCapability>,
+        cycle_root: SourceTextModule,
+        error: Option<JsError>,
+    },
+}
+
+// SAFETY: This must be synced with `Status` to mark any new data added that needs to be traced.
+// This implementation is necessary to be able to transition from one state to another by destructuring,
+// which saves us some unnecessary clones.
+unsafe impl Trace for Status {
+    custom_trace!(this, {
+        match this {
+            Status::Unlinked | Status::Linking { .. } | Status::Linked { .. } => {}
```

Comment:
Couldn't we use the `Trace` proc macro with `#[unsafe_ignore_trace]` attribute

Reply:
Not really. The `Trace` proc macro adds a `Drop` implementation to disencourage implementing `Drop` on `Trace` implementors. However, implementing `Drop` disallows destructuring, which is pretty useful to have for state types like `Status`.